src/main/kotlin/LocalEvaluationClient.kt

```diff
@@ -67,15 +94,108 @@ class LocalEvaluationClient internal constructor(
         if (sortedFlagConfigs.isEmpty()) {
             return mapOf()
         }
-        val evaluationResults = evaluation.evaluate(user.toEvaluationContext(), sortedFlagConfigs)
-        Logger.d("evaluate - user=$user, result=$evaluationResults")
+        val enrichedUser = enrichUser(user, sortedFlagConfigs)
+        val evaluationResults = wrapMetrics(
+            metric = metrics::onEvaluation,
+            failure = metrics::onEvaluationFailure,
+        ) {
+            evaluation.evaluate(enrichedUser.toEvaluationContext(), sortedFlagConfigs)
+        }
+        Logger.d("evaluate - user=$enrichedUser, result=$evaluationResults")
         val variants = evaluationResults.toVariants()
         assignmentService?.track(Assignment(user, variants))
         return variants
     }
+
+    private fun enrichUser(user: ExperimentUser, flagConfigs: List<EvaluationFlag>): ExperimentUser {
+        val groupedCohortIds = flagConfigs.getGroupedCohortIds()
+        if (cohortStorage == null) {
+            if (groupedCohortIds.isNotEmpty()) {
+                Logger.e("Local evaluation flags target cohorts but cohort targeting is not configured.")
+            }
+            return user
+        }
+        return user.copyToBuilder().apply {
+            val userCohortsIds = groupedCohortIds[USER_GROUP_TYPE]
+            if (!userCohortsIds.isNullOrEmpty() && user.userId != null) {
+                cohortIds(cohortStorage.getCohortsForUser(user.userId, userCohortsIds))
+            }
+            if (user.groups != null) {
+                for (group in user.groups) {
+                    val groupType = group.key
+                    val groupName = group.value.firstOrNull() ?: continue
+                    val cohortIds = groupedCohortIds[groupType]
+                    if (cohortIds.isNullOrEmpty()) {
+                        continue
+                    }
+                    groupCohortIds(
+                        groupType,
+                        groupName,
+                        cohortStorage.getCohortsForGroup(groupType, groupName, cohortIds)
+                    )
+                }
+            }
+        }.build()
+    }
+}
+
+private fun getCohortDownloadApi(config: LocalEvaluationConfig, httpClient: OkHttpClient): CohortApi? {
+    return if (config.cohortSyncConfig != null) {
+        DynamicCohortApi(
+            apiKey = config.cohortSyncConfig.apiKey,
+            secretKey = config.cohortSyncConfig.secretKey,
+            maxCohortSize = config.cohortSyncConfig.maxCohortSize,
+            serverUrl = getCohortServerUrl(config),
+            proxyUrl = getProxyUrl(config),
+            httpClient = httpClient,
+        )
+    } else {
+        null
+    }
+}
+
+private fun getServerUrl(config: LocalEvaluationConfig): HttpUrl {
+    return if (config.serverZone == LocalEvaluationConfig.Defaults.SERVER_ZONE) {
+        config.serverUrl.toHttpUrl()
+    } else {
+        when (config.serverZone) {
```

Comment:
So if an EU user wants to set a custom URL, they should skip setting serverZone?

Reply:
You're right switching up logic